### DISCUSSION TITLE
How to build custom endpoints in dev env without re-building an extension each time

### DISCUSSION THREAD

**Discussion by darkylmnx** [directus/directus#17407] [Archive]
Reading the docs, I found out that all custom endpoints must be built as an extension,
even when it's for local user.

The issue is, when I'm in dev environment, I'd like to not be forced to re-build the extension each time.
It's really creating an unpleasant DX for me.

So I have 2 questions :
- Is there a way to not have to create an extension for custom endpoints and just register them locally directly?
- If there isn't, is there a way to not have to rebuild the extension each time I do a modification?

Thanks!

**Comment by JonathanSchndr** [comment]
The reason for this is so that your extension is converted from ESM to Node, which Directus can interpret and also includes all dependencies automatically. You can simply append a --watch, then your extension will be recompiled with every change.

**Reply by darkylmnx** [reply]
I saw that but isn't recent versions of node ESM compatible?
Why is this still necessary?

Expecially for local extensions, specifically for just adding an endpoint, this really feels like an overhead.

Is there not just a way to create the custom endpoint in a js file in `/extensions/endpoints/custom.js` and have `directus start --watch` watch the files of the project?

Or is that what you meant here, and not `directus-extension build --watch` ?

**Reply by JonathanSchndr** [reply]
That is my watch config, I hope this helps you:

```
"scripts": {
    "build": "directus-extension build",
    "watch": "npm-watch"
  },
  "dependencies": {
    "npm-watch": "^0.11.0"
  }
```

**Reply by darkylmnx** [reply]
Okay so as I thought, you meant watching the extension.

Do you think it's possible to just create the custom endpoint in a js file in `/extensions/endpoints/custom.js` and have `directus start --watch` watch the files of the project?

@rijkvanzanten any thoughts?

**Reply by paescuj** [reply]
The main reason for the build process of extensions is actually to include (bundle) external dependencies into the final output.
If you don't rely on any external dependencies you're free to just create and edit the file directly at `/extensions/endpoints/my-custom-extension/index.js`.
If you want to use ESM, you must name the file `index.mjs` (see https://docs.directus.io/extensions/creating-extensions.html#configuring-the-cli).
For `directus start --watch` it doesn't matter how the file has been created, it will be reloaded on changes anyway.

Please don't randomly ping other users. Thank you!

**Reply by darkylmnx** [reply]
Sorry for the ping. The main reason was that what you just explained isn't explained anywhere in the docs.

Even the link you just gave doesn't add contexte, because the docs says:

> To install an extension locally, you have to move the output from the dist/ folder into your project's ./extensions/<extension-folder>/<extension-name>/ folder. <extension-folder> has to be replaced by the extension type in plural form (e.g. interfaces). <extension-name> should be replaced with the name of your extension.

Which in no way one could guess that having a file directly into the extension would work.

One would also assume that Directus has global/internal way to bundle things on the fly during development.

In any case, thanks for the answer.

**Reply by paescuj** [reply]
> Sorry for the ping. The main reason was that what you just explained isn't explained anywhere in the docs.

No problem!

> Which in no way one could guess that having a file directly into the extension would work.

This is not something which is usually requested nor the official / recommended way to build an extension so that's why it's not mentioned explicitly in the docs.

> In any case, thanks for the answer.

You're welcome!

**Reply by claytongulick** [reply]
We implement all of our endpoint extensions this way, just with a direct index.js file and normal commonjs node code. No build step. We have a ton of endpoint extensions implemented like this, running in production.

FWIW, because I'm also allergic to transpilation, and really didn't like the DX of the UI extensions or being limited to Vue, we actually at one point were operating on a forked version that added a /app path that would allow us to directly include arbitrary js without a build into the extensions - so we had vanilla web components loading and running as interfaces and all sorts of stuff. Worked great.

We don't do that anymore because we've since developed app UIs for a more robust experience, but it got us off the ground quickly.

**Reply by darkylmnx** [reply]
Thanks for your feedback on that!

**Reply by martin-braun** [reply]
I stumbled upon this thread. I remember doing things the old fashioned way and simply putting index.js into `/extensions/endpoints/abc/` and it would be available under `/custom/abc/<subroute>`. Picking up a new version, it doesn't work anymore and the docs only instruct to build an extension. But what if the extension I want to build is tied to my project so tightly that I don't want to move my endpoints off the project? Also the whole "move contents of dist to extensions" part annoys me. 

After a lot of trial and error I figured that you can do what I want to do, simply `cd extensions` and then create your extension using `npx create-directus-extension@latest`. I name the extension how I want the route to be and since I just have a few custom endpoints in my project, I can just name the extension `api` and start from there. 

Directus seem to search for the `dist` folder in your `extension/<my-extension>` folder and loads it properly. So, to put things together, my top `package.json` needs to cascade down to ensure that my custom endpoints build. I don't care if this is recommend or not, I need something simple to start, so I modified my package.json like so:

```json
  "scripts": {
    "start": "directus database migrate:latest && directus start",
    "build:api": "( cd extensions/api && directus-extension build )",
    "build": "run-p build:api",
    "dev:api": "( cd extensions/api && directus-extension build -w --no-minify )",
    "dev": "run-p dev:api start"
  },
```

`run-p` comes from `npm i npm-run-all -D`. It works pretty much.

**Reply by vongoethe** [reply]
> I stumbled upon this thread. I remember doing things the old fashioned way and simply putting index.js into `/extensions/endpoints/abc/` and it would be available under `/custom/abc/<subroute>`. Picking up a new version, it doesn't work anymore and the docs only instruct to build an extension. But what if the extension I want to build is tied to my project so tightly that I don't want to move my endpoints off the project? Also the whole "move contents of dist to extensions" part annoys me.
> 
> After a lot of trial and error I figured that you can do what I want to do, simply `cd extensions` and then create your extension using `npx create-directus-extension@latest`. I name the extension how I want the route to be and since I just have a few custom endpoints in my project, I can just name the extension `api` and start from there.
> 
> Directus seem to search for the `dist` folder in your `extension/<my-extension>` folder and loads it properly. So, to put things together, my top `package.json` needs to cascade down to ensure that my custom endpoints build. I don't care if this is recommend or not, I need something simple to start, so I modified my package.json like so:
> 
> ```json
>   "scripts": {
>     "start": "directus database migrate:latest && directus start",
>     "build:api": "( cd extensions/api && directus-extension build )",
>     "build": "run-p build:api",
>     "dev:api": "( cd extensions/api && directus-extension build -w --no-minify )",
>     "dev": "run-p dev:api start"
>   },
> ```
> 
> `run-p` comes from `npm i npm-run-all -D`. It works pretty much.

This is useful. Thank you.
